File: Evaluate Reverse Polish Notation (RPN) - Leetcode 150.py
from math import ceil, floor
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stk = []
        for t in tokens:
            if t in "+-*/":
                b, a = stk.pop(), stk.pop()

                if t == "+":
                    stk.append(a + b)
                elif t == "-":
                    stk.append(a - b)
                elif t == "*":
                    stk.append(a * b)
                else:
                    division = a / b
                    if division < 0:
                        stk.append(ceil(division))
                    else:
                        stk.append(floor(division))
            else:
                stk.append(int(t))

        return stk[0]

# Time Complexity: O(n)
# Space Complexity: O(n)

File: Evaluate Reverse Polish Notation (RPN) - Leetcode 150/Evaluate Reverse Polish Notation (RPN) - Leetcode 150.cpp
#include <vector>
#include <string>
#include <stack>
#include <cmath> // For std::ceil and std::floor
using namespace std;

/**
 * @param tokens: A vector of strings representing the RPN expression
 * @return: The result of the RPN expression
 */
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> stk;

        for (const auto& t : tokens) {
            if (t == "+" || t == "-" || t == "*" || t == "/") {
                int b = stk.top(); stk.pop();
                int a = stk.top(); stk.pop();
                
                if (t == "+") stk.push(a + b);
                else if (t == "-") stk.push(a - b);
                else if (t == "*") stk.push(a * b);
                else {
                    double division = static_cast<double>(a) / b;
                    stk.push(division < 0 ? static_cast<int>(ceil(division)) : static_cast<int>(floor(division)));
                }
            } else {
                stk.push(stoi(t));
            }
        }
        
        return stk.top();
    }
};

// Time Complexity: O(n)
// Space Complexity: O(n)
